node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,411 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metered variant of oneshot channels to be able to extract delays caused by delayed responses.
+
+use std::{
+	ops::Deref,
+	pin::Pin,
+	task::{Context, Poll},
+	time::{Duration, Instant},
+};
+
+use futures::{
+	channel::oneshot::{self, Canceled, Cancellation},
+	future::{Fuse, FusedFuture},
+	prelude::*,
+};
+use futures_timer::Delay;
+
+/// Provides the reason for termination.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Reason {
+	Completion = 1,
+	Cancellation = 2,
+	HardTimeout = 3,
+}
+
+/// Obtained measurements by the `Receiver` side of the `MeteredOneshot`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Measurements {
+	/// Duration between first poll and polling termination.
+	first_poll_till_end: Duration,
+	/// Duration starting with creation until polling termination.
+	creation_till_end: Duration,
+	/// Reason for resolving the future.
+	reason: Reason,
+}
+
+impl Measurements {
+	/// Obtain the duration of a finished or canceled
+	/// `oneshot` channel.
+	pub fn duration_since_first_poll(&self) -> &Duration {
+		&self.first_poll_till_end
+	}
+
+	/// Obtain the duration of a finished or canceled
+	/// `oneshot` channel.
+	pub fn duration_since_creation(&self) -> &Duration {
+		&self.creation_till_end
+	}
+
+	/// Obtain the reason to the channel termination.
+	pub fn reason(&self) -> &Reason {
+		&self.reason
+	}
+}
+
+/// Create a new pair of `OneshotMetered{Sender,Receiver}`.
+pub fn channel<T>(
+	name: &'static str,
+	soft_timeout: Duration,
+	hard_timeout: Duration,
+) -> (MeteredSender<T>, MeteredReceiver<T>) {
+	let (tx, rx) = oneshot::channel();
+
+	(
+		MeteredSender { name, inner: tx },
+		MeteredReceiver {
+			name,
+			inner: rx,
+			soft_timeout,
+			hard_timeout,
+			soft_timeout_fut: None,
+			hard_timeout_fut: None,
+			first_poll_timestamp: None,
+			creation_timestamp: Instant::now(),
+		},
+	)
+}
+
+#[allow(missing_docs)]
```

Comment:
why is this `allow(missing_docs)`? the `measurements` function is also without docs.

Reply:
This in particular has missing docs, because iff the error is not self describing, there is something wrong already. The error annotation should be enough. This is a pattern we have in more places throughout the codebase, limited to `Error` types.